File: src/main/java/ar/com/flow/bankaccount/transaction/Step.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.bankaccount.BankAccount;
import ar.com.flow.bankaccount.transaction.receipt.Receipt;
import ar.com.flow.bankaccount.transaction.receipt.ReceiptStamper;
import ar.com.flow.money.Money;
import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
public class Step {
    private final BankAccount account;
    private final ReceiptStamper receiptStamper;

    Receipt execute(Money amount) {
        return register(receiptStamper.print(amount));
    }

    Receipt undo(Money amount) {
        return register(receiptStamper.scratch(amount));
    }

    private Receipt register(Receipt receipt) {
        account.addReceipt(receipt);
        return receipt;
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/transaction/Step.kt
package ar.com.flow.bankaccount.transaction

import ar.com.flow.bankaccount.BankAccount
import ar.com.flow.bankaccount.transaction.receipt.Receipt
import ar.com.flow.bankaccount.transaction.receipt.ReceiptPrinter
import ar.com.flow.money.Money

class Step(
    private val account: BankAccount,
    private val receiptPrinter: ReceiptPrinter
) {
    fun execute(amount: Money): Receipt {
        return register(receiptPrinter.print(amount))
    }

    fun undo(amount: Money): Receipt {
        return register(receiptPrinter.scratch(amount))
    }

    private fun register(receipt: Receipt): Receipt {
        account.addReceipt(receipt)
        return receipt
    }
}